### ISSUE TITLE
Identical devices get detected as different types

### ISSUE THREAD

**Issue by novanarwhal** [make-all/tuya-local#1959]
**Describe the bug**
I am adding 8 fans (detected as Skyfan DC fans). Two of the devices are being detected as a light.
**To Reproduce**
Adding the fan using the cloud assisted configuration

**Expected behavior**
All fans to be added as Skyfan DC fans.

**Additional context**
Attached diagnostics from correct and incorrectly detected fans.

Is there any way to manually edit devices using the configuration from a correct fan into the incorrect fan?

![Screenshot 2024-05-28 113822](https://github.com/make-all/tuya-local/assets/170793470/87a8efa2-c8e0-4402-8062-3af3189730e1)
[tuya_local-2bc2c9e9f69fb1fd99fcdee686405e8b-Guest fan-6c5aad47d776fb3b90314cb93661ebd9.json](https://github.com/make-all/tuya-local/files/15462885/tuya_local-2bc2c9e9f69fb1fd99fcdee686405e8b-Guest.fan-6c5aad47d776fb3b90314cb93661ebd9.json)
[tuya_local-5e2146fc35846b11c844e35c96a514f9-RGBCW light rumpus-8c9d7023e1eb944b874a5314c7f8890e.json](https://github.com/make-all/tuya-local/files/15462887/tuya_local-5e2146fc35846b11c844e35c96a514f9-RGBCW.light.rumpus-8c9d7023e1eb944b874a5314c7f8890e.json)
![Screenshot 2024-05-28 115131](https://github.com/make-all/tuya-local/assets/170793470/53d30813-75e9-4569-b3e9-75a51b8ec8e6)

**Comment by make-all** [comment]
Changing the device type after installation changes the entities which are present, messing up the entity registry. So the only supported way to do it is to remove and re-add the device.

**Comment by make-all** [comment]
You should have been offered the fan config when you installed the devices.  It seems you just clicked through with the default option selected without checking it first.

**Comment by novanarwhal** [comment]
> You should have been offered the fan config when you installed the devices. It seems you just clicked through with the default option selected without checking it first.
![Screenshot 2024-05-28 115131](https://github.com/make-all/tuya-local/assets/170793470/1fb7cd9e-e584-4e83-b36d-fea50eae52c3)


This is the option list provided when trying to setup the problem fans. No fan option available.

**Comment by make-all** [comment]
Sorry, I read these configs the wrong way around, thinking that the misdetected one had an extra dp, when it was actually a missing dp for the preset.

**Comment by novanarwhal** [comment]
> > You should have been offered the fan config when you installed the devices. It seems you just clicked through with the default option selected without checking it first.
> > ![Screenshot 2024-05-28 115131](https://private-user-images.githubusercontent.com/170793470/334243522-1fb7cd9e-e584-4e83-b36d-fea50eae52c3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY4NjE1MzIsIm5iZiI6MTcxNjg2MTIzMiwicGF0aCI6Ii8xNzA3OTM0NzAvMzM0MjQzNTIyLTFmYjdjZDllLWU1ODQtNGU4My1iMzZkLWZlYTUwZWFlNTJjMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyOFQwMTUzNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYTIxN2NjOGNmMTljMjFiMDNhZThiYjJlNmEwNTNlNmZmYTE3ZTI4MzRiMzhkMjE0MDE2MjBjOGVjMGM3YjRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.64BwoUMgrdtYEe-FSAFCOb_oKeIFz2FoveuAsQwy-94)
> 
> This is the option list provided when trying to setup the problem fans. No fan option available.

For reference this is one that detects properly.
![Screenshot 2024-05-28 115519](https://github.com/make-all/tuya-local/assets/170793470/ca2c3c0f-c7e5-4188-b548-ceee5710bf58)

**Comment by novanarwhal** [comment]
Thanks for looking in at this. Anyway to temporarily address this or edit it before a fix flows through?

**Comment by make-all** [comment]
Add `optional: true` to the preset_mode dp config.

**Comment by novanarwhal** [comment]
> Add `optional: true` to the preset_mode dp config.

I do apologise -  you have blown right past my level of knowledge here. Can you tell me where I go to add this or point to a guide?

**Comment by make-all** [comment]
Find the file skyfan_dc_fan.yaml.  In that file, search for preset_mode.  On the line after that, add a new line with same indentation that says `optional: true`

**Comment by novanarwhal** [comment]
![Screenshot 2024-05-28 171343](https://github.com/make-all/tuya-local/assets/170793470/49ad14f4-862d-4089-83d7-f12d85a94219)

Unfortunately this change does not enable me to chose the fan option at the setup stage.

**Comment by novanarwhal** [comment]
I have managed to get it work. I added the optional line to a different dps section and it started getting detected.
![Screenshot 2024-05-28 173026](https://github.com/make-all/tuya-local/assets/170793470/0191ee44-a09d-4e01-86d5-4d9ead6ea53c)

Thank you for your help.

**Comment by make-all** [comment]
Strange. The device must randomly fail to report either speed or preset (maybe depending on which was last used).

**Comment by novanarwhal** [comment]
If there is anything I can do to assist with identifying a cause of
outcome, let me know.

**Comment by novanarwhal** [comment]
> Strange. The device must randomly fail to report either speed or preset (maybe depending on which was last used).

I played a bit more. It would seem that it entirely depends on when it was last used. I removed the `optional: true` line from the yaml and then turned a fan on (via remote, not isolator), set a speed and a timer that previously didnt work with tuya-local and all of a sudden it started getting detected correctly.

So the best solution to this problem is ensure that the device has been activated and certain functions used.
